### PR TITLE
Issue 6256 - nsslapd-numlisteners limit is not enforced

### PR DESCRIPTION
**Pull request by ssidhaye**
Description: Add a test to check if nsslapd-numlisteners value
can be set higher than 4

Relates: https://github.com/389ds/389-ds-base/issues/6256

Reviewed by: droideck

### CODE REVIEW

File: dirsrvtests/tests/suites/config/config_test.py [hunk 1]
```diff
@@ -741,6 +742,57 @@ def test_lmdb_config(create_lmdb_instance):
...
+    topo.standalone.config.set('nsslapd-numlisteners', '4')
+
+    # Check nsslapd-numlisteners value is set to 4
+    assert topo.standalone.config.get_attr_val_utf8('nsslapd-numlisteners') == '4'
+
+    # Check instance dse.ldif value is set to 5
+    inst = topo.standalone
+    dse_ldif = DSEldif(inst)
+    numlisteners = dse_ldif.get("cn='Directory Manager'", 'nsslapd-numlisteners')
+    assert numlisteners[0] == '5'
```
Comment (droideck) [inline dirsrvtests/tests/suites/config/config_test.py]: When I run the test, I get this failure:

```
        # Check instance dse.ldif value is set to 5
        inst = topo.standalone
        dse_ldif = DSEldif(inst)
        numlisteners = dse_ldif.get("cn='Directory Manager'", 'nsslapd-numlisteners')
>       assert numlisteners[0] == '5'
E       TypeError: 'NoneType' object is not subscriptable
```

Comment (droideck) [reply dirsrvtests/tests/suites/config/config_test.py]: Also, I wonder why `nsslapd-numlisteners` should be `5` when in `dse.ldif` but `4` via search?

**Review by droideck — CHANGES_REQUESTED** [review]

File: dirsrvtests/tests/suites/config/config_test.py (hunk 1 above)
Comment (jchapma) [reply dirsrvtests/tests/suites/config/config_test.py]: Hey @droideck 
There is a "feature" in the numlistener config code that allows an invalid ```nsslapd-numlisteners``` attribute value be written to dse.ldif before the config code gets a chance to sanitise it.

File: dirsrvtests/tests/suites/config/config_test.py [hunk 2]
```diff
@@ -741,6 +742,57 @@ def test_lmdb_config(create_lmdb_instance):
...
+    # Set nsslapd-numlisteners value to 4
+    topo.standalone.config.set('nsslapd-numlisteners', '4')
+
+    # Check nsslapd-numlisteners value is set to 4
+    assert topo.standalone.config.get_attr_val_utf8('nsslapd-numlisteners') == '4'
+
+    # Check instance dse.ldif value is set to 5
+    inst = topo.standalone
+    dse_ldif = DSEldif(inst)
+    numlisteners = dse_ldif.get("cn='Directory Manager'", 'nsslapd-numlisteners')
```
Comment (jchapma) [inline dirsrvtests/tests/suites/config/config_test.py]: I think the reason we get None here is down to the entrydn value in the call to dse_ldif.get(). The ```nsslapd-numlisteners``` attribute and value can be found in "cn=config".

File: dirsrvtests/tests/suites/config/config_test.py [hunk 3]
```diff
@@ -741,6 +742,57 @@ def test_lmdb_config(create_lmdb_instance):
...
+        6. nsslapd-numlisteners value should be successfully set to 5
+        7. nsslapd-numlisteners value is indeed set to 5
+    """
+    # Check default value for nsslapd-numlisteners is 1
+    assert topo.standalone.config.get_attr_val_utf8('nsslapd-numlisteners') == '1'
+
+    # Set nsslapd-numlisteners value to 4
+    topo.standalone.config.set('nsslapd-numlisteners', '4')
+
+    # Check nsslapd-numlisteners value is set to 4
```
Comment (jchapma) [inline dirsrvtests/tests/suites/config/config_test.py:774]: When we "set" the value for ````nsslapd-numlisteners``` we need to restart for the change to take effect, but maybe in this case where its just the config we are dealing with, we dont ?

**Review by jchapma — COMMENTED** [review]
Hey @ssidhaye 

There is a bit of overlap between this PR and [6257](https://github.com/389ds/389-ds-base/pull/6257). 6257 contains a test similar to yours and a fix for this issue. Now that I think about it, is it better to push test cases for a particular fix in a separate PR. If that is the case then I can push the fix after you push this test case. Whatever you think.

File: dirsrvtests/tests/suites/config/config_test.py (hunk 2 above)
Comment (jchapma) [reply dirsrvtests/tests/suites/config/config_test.py]: But I do prefer the use of DSEldif() over the way I handle dse.ldif in 6257

File: dirsrvtests/tests/suites/config/config_test.py [hunk 4]
```diff
@@ -741,6 +742,57 @@ def test_lmdb_config(create_lmdb_instance):
...
+    inst = topo.standalone
+    dse_ldif = DSEldif(inst)
+    numlisteners = dse_ldif.get("cn='Directory Manager'", 'nsslapd-numlisteners')
+    assert numlisteners[0] == '5'
+
+    # Restart instance
+    topo.standalone.restart()
+
+    # Check nsslapd-numlisteners value is set to 5
+    assert topo.standalone.config.get_attr_val_utf8('nsslapd-numlisteners') == '5'
```
Comment (jchapma) [inline dirsrvtests/tests/suites/config/config_test.py]: The max value for the ```nsslapd-numlisteners``` attr is 4, so if we get 5 we are in trouble

File: dirsrvtests/tests/suites/config/config_test.py (hunk 1 above)
Comment (ssidhaye) [reply dirsrvtests/tests/suites/config/config_test.py]: I forgot to replace  `Directory Manager` with `DN_CONFIG`.
Sorry about the check for 5. I was trying out a few things and did not change it back to 4.
Fixed it

File: dirsrvtests/tests/suites/config/config_test.py (hunk 2 above)
Comment (ssidhaye) [reply dirsrvtests/tests/suites/config/config_test.py]: Yes I changed it to use `DN_CONFIG`

File: dirsrvtests/tests/suites/config/config_test.py (hunk 4 above)
Comment (ssidhaye) [reply dirsrvtests/tests/suites/config/config_test.py]: Yes, fixed it. I misread it earlier.

**Review by droideck — APPROVED** [review]
LGTM

File: dirsrvtests/tests/suites/config/config_test.py (hunk 3 above)
Comment (ssidhaye) [reply dirsrvtests/tests/suites/config/config_test.py:774]: Yes, we don't need to restart

**Pull request merged**